cfy_manager/components/base_component/base_component.py

```diff
@@ -82,3 +88,90 @@ def validate_dependencies(self):
             self.logger.debug(
                 'All prerequisites for {class_name} are met'.format(
                     class_name=self.__class__.__name__))
+
+    def use_supplied_certificates(self, cert_destination, key_destination,
+                                  ca_destination,
+                                  owner=CLOUDIFY_USER, group=CLOUDIFY_GROUP):
+        """
+            Use user-supplied certificates, after checking they're not
+            broken.
+        """
+        cert_src, key_src, ca_src, key_pass = check_certificates(
+            self.component_name,
+        )
+
+        if not any([cert_src, key_src, ca_src, key_pass]):
+            # No certificates supplied, so not using them
+            self.logger.debug('No user-supplied certificates were present.')
+            return False
+
+        if key_pass:
+            # TODO: Strip password from key
+            raise ValidationError('Key must not have password.')
+
+        # Put the files in the correct place
+        self.logger.info('Ensuring files are in correct locations.')
+        for path in cert_destination, key_destination, ca_destination:
+            directory = os.path.dirname(path)
+            if not os.path.exists(directory):
+                self.logger.info(
+                    'Ensuring certificate destination {directory} exists '
+                    'and is accessible.'.format(directory=directory)
+                )
+                sudo(['mkdir', '-p', directory])
+                sudo(['chmod', 'o+x', directory])
+            if not os.path.isdir(directory):
+                raise ValidationError(
+                    'Certificate destination directory {directory} exists '
+                    'but is not a directory!'.format(directory=directory)
+                )
+
+        if cert_src != cert_destination:
```

Comment:
Feels like this whole section (#L129-#L160) can be simplified using an external/internal logging and file moving function?

Reply:
I was considering that, but:
1. We don't have one yet (do we?).
2. There are only two of these that are basically exactly the same. The CA cert one differs and working on a rule of three (only factoring out when doing the same thing three times) I figured it wasn't worth doing.

I could factor it out though- worth it or not?

Reply:
But even in the rule-of-3 you still have the "log and move file" commands which can be factored out.

Reply:
That's true, I'll add something to the utils.files

Reply:
Oh, wait... we have it. It's in common, not files.

Reply:
Shifted to using the methods from utils.common